### PR TITLE
add profile case transformations to "aws-sso-util configure populate"

### PR DESCRIPTION
**Pull request by oshaughnessy**
which provides a way to evaluate Python code against profile names that are built from components, giving a CLI middle ground between trimming regexes and invoking a separate script

### CODE REVIEW

**Comment by benkehoe** [comment]
Sorry, but I can't accept this. I understand the pain point, but I'm not willing to put `eval` in the code. If you have other ideas about how to provide more options or more flexible profile name transformations, I'm definitely open to it, but running arbitrary code from a command line argument isn't a good practice.

**Comment by oshaughnessy** [comment]
Understood. That crossed my mind as well :^)  It's safe enough in my use case, but I understand. I'll come up with an alternative plan. I'd rather do something that can get merged into your baseline than maintain a fork.

**Comment by benkehoe** [comment]
Do you have a sense of what transforms would be useful? You've got lowercasing in the example, but that's only one. Another potential avenue is allowing arguments to the transform script to be passed through the `configure populate` command arguments.

**Comment by oshaughnessy** [comment]
Hi, Ben. Thanks for following up!

My use case is really just lowercasing, but I thought something more flexible would make for a better feature. I just updated my branch to do something long those lines, so with it now, you can specify `--transform-profile-name lower` or select from a variety of other simple choices (or do more than one, although i suspect any utility there is rarely going to add value):

    * `alnum`
    * `ascii`
    * `capitalize`
    * `casefold`
    * `lower`
    * `swapcase`
    * `title`
    * `upper`

**Comment by benkehoe** [comment]
This is more reasonable. I'll have to think about it some. I'm preparing a release now, so it would go in the next one.

**Comment by oshaughnessy** [comment]
Sure. I’m sure you’re more than capable, but let me know if you’d prefer it a bit different. Happy to adjust it. I just wanted a nice way to lowercase my profiles at a new job without writing my own solution again. Quite pleased you’ve share this — I hadn’t come across it until just recently. 

One thing I debated was aligning the arg name and method with the trim features, or aligning it with the safe-account-names feature. The method I chose, replicating how you implemented the trim features, lets us chain transforms, so there’s a future for more powerful capabilities later (even though the current set of choices don’t do much to take advantage of that).

**Comment by benkehoe** [comment]
I'll keep pondering on it. It's a worthwhile request, the profile name process means you have to construct the entire profile name yourself, as an escape hatch it's a lot of work. Here's another idea: what if there was an `--profile-name-transform-process` that piped the full name, as determined by the other arguments, to stdin? So like you could use `--profile-name-transform-process "tr '[:upper:]' '[:lower:]'"` or `--profile-name-transform-process "python -c 'import sys; sys.stdout.write(sys.stdin.read().lower())'"`

**Comment by oshaughnessy** [comment]
Yeah, that could be a nice solution, too. That sort of thing is actually where my unix-brain went first... "a way to pass the profile name through sed would be really convenient," and `tr` is a nice refinement on that. My love of running things through pipes to UNIX utilities really likes this kind of solution, and I think you'll end up with different use cases than I'm solving w/my PR. I kinda like having both, although that feels a little like feature creep. `--transform-profile-name lower` is easier for people to call, it's lighter weight because it's just calling `str.lower()`, and it's going to look better in somebody's README when they tell people at their org how to set up their profiles. But, again, the UNIX hacker in me really likes the idea of having a way to run profile names through a subprocess.

**Comment by benkehoe** [comment]
Ok: 
* Let's call this `--account-name-case` and `--role-name-case`, and apply them separately. Each can be provided only once.
* Keep only `lower`, `upper`, `title`, `capitalize`, and `casefold`.

`--safe-account-names` already does what `alnum`/`ascii` does, they aren't needed, and roles are already only alphanumeric. Nobody is going to use `swapcase`.

**Comment by oshaughnessy** [comment]
> Nobody is going to use swapcase.

Nobody but the chaos monkeys. I just put that one in for fun ;-)

**Comment by oshaughnessy** [comment]
Pushed an update. New help message includes this:

```
  --account-name-case [capitalize|casefold|lower|upper]
                                  Method to change the case of the account
                                  name

  --role-name-case [capitalize|casefold|lower|upper]
                                  Method to change the case of the role name
``

File: docs/configure.md [hunk 1]
```diff
@@ -236,3 +236,26 @@ region_str = "" if region_index == 0 else sep + short_region_name
 print(account_name + sep + role_name + region_str)
 ```
 If this was stored as `profile_formatter.py`, it could be used as `--profile-name-process "python profile_formatter.py"`
+
+### Change the case of profile names
+
+If you don't want to run (and maintain) a process completely outside of the
```
Comment (benkehoe) [inline docs/configure.md:310]: Because this isn't arbitrary transformations, I don't think this needs to mention profile name process. Let's also move this section to after trim and before process

**Comment by oshaughnessy** [comment]
In my last commit, I added the output from `aws-sso-util configure populate --help` to the configure doc, too. I was looking for something like that earlier and couldn't find it anywhere other than my terminal. I thought it would be helpful to have it online.

File: docs/configure.md [hunk 2]
```diff
@@ -236,3 +236,26 @@ region_str = "" if region_index == 0 else sep + short_region_name
...
+`--profile-name-process`, an option between trimming account & role names and
+executing a separate process is available with `--account-name-case` and
+`--role-name-case`. Use these to change the generated profile names after
+assembly and trimming.
+
+Case modifications include:
```
Comment (benkehoe) [inline docs/configure.md:316]: "The values for these options are as follows, and correspond to the Python string methods of the same names:"
And then link each one to the string method docs

File: docs/configure.md [hunk 3]
```diff
@@ -236,3 +236,26 @@ region_str = "" if region_index == 0 else sep + short_region_name
...
+    * `capitalize`
+    * `casefold`
+    * `lower`
+    * `upper`
+
+For example, to lowercase both the account name and role name, add these options:
+
+    --account-name-case lower --role-name-case lower
+
+If you want to see what it does, try a dry run:
```
Comment (benkehoe) [inline docs/configure.md:327]: I don't think we need the dry run example.

File: cli/src/aws_sso_util/populate_profiles.py [hunk 4]
```diff
@@ -149,6 +149,23 @@ def trim_formatter(i, n, **kwargs):
         return formatter(i, n, **kwargs)
     return trim_formatter
 
+def get_field_name_case_formatter(field, transform, formatter):
+    def case_formatter(i, n, **kwargs):
+        LOGGER.debug("evaluating case transform(%s)", transform)
```
Comment (benkehoe) [inline cli/src/aws_sso_util/populate_profiles.py:154]: Let's remove the logger statements in here, they should either be in all the formatters or none, so it would be a separate PR

File: cli/src/aws_sso_util/populate_profiles.py [hunk 5]
```diff
@@ -244,6 +265,10 @@ def populate_profiles(
         profile_name_formatter = get_formatter(profile_name_include_region, region_format, no_region_format)
         if profile_name_trim_account_name_patterns or profile_name_trim_role_name_patterns:
             profile_name_formatter = get_trim_formatter(profile_name_trim_account_name_patterns, profile_name_trim_role_name_patterns, profile_name_formatter)
+        if profile_name_account_name_case_transform:
```
Comment (benkehoe) [inline cli/src/aws_sso_util/populate_profiles.py:268]: Let's make this work like `get_trim_formatter`, where it operates on both fields at once, rather than having the general-purpose field transform as it is. make it `get_case_formatter(profile_name_account_name_case_transform, profile_name_role_name_case_transform, profile_name_formatter)`

File: docs/configure.md [hunk 6]
```diff
@@ -157,6 +157,74 @@ aws s3 ls --profile my-sso-profile
 `aws-sso-util configure populate` allows you to configure profiles for all the access you have through AWS SSO.
 You specify one or more regions, and a profile is created for every account, role, and region you have access to through AWS SSO (note that if access to a region is prohibited by an IAM policy, this does not suppress creation of the profile).
 
+## Options
```
Comment (benkehoe) [inline docs/configure.md:160]: Pull this section out. It's a good point that documentation of the options is missing but I'd rather have it formatted the same as other commands. I can do that after.

**Comment by benkehoe** [comment]
Add yourself to `CONTRIBUTORS.md`

File: docs/configure.md [hunk 7]
```diff
@@ -206,6 +274,27 @@ A useful piece of syntax for this is lookahead/lookbehind assertions.
...
+    * [`capitalize`](https://docs.python.org/3/library/stdtypes.html#str.capitalize)
+    * [`casefold`](https://docs.python.org/3/library/stdtypes.html#str.casefold)
+    * [`lower`](https://docs.python.org/3/library/stdtypes.html#str.lower)
+    * [`upper`](https://docs.python.org/3/library/stdtypes.html#str.upper)
+
+For example, to lowercase both the account name and role name, add these options:
+
+    --account-name-case lower --role-name-case lower
+
+If you want to see what it does, try a dry run:
```
Comment (benkehoe) [inline docs/configure.md:294]: Remove the dry run example, we cover it elsewhere

File: cli/src/aws_sso_util/populate_profiles.py [hunk 8]
```diff
@@ -149,6 +149,23 @@ def trim_formatter(i, n, **kwargs):
         return formatter(i, n, **kwargs)
     return trim_formatter
 
+def get_field_name_case_formatter(field, transform, formatter):
+    def case_formatter(i, n, **kwargs):
+        LOGGER.debug("evaluating case transform(%s)", transform)
+        LOGGER.debug("* in  = %s", kwargs[field])
+        if transform == 'capitalize':
```
Comment (benkehoe) [inline cli/src/aws_sso_util/populate_profiles.py:156]: With the change to this function to operate directly on the fields, I think this looks more like:
```python
for field, transform in [("account_name", account_name_case), ("role_name", role_name_case)]:
    if not transform:
        continue
    if transform == "capitalize":
        kwargs[field] = kwargs[field].capitalize()
    # etc.
```
Make it more direct rather than the generic transform application happening here.

**Review by benkehoe — CHANGES_REQUESTED** [review]

**Comment by oshaughnessy** [comment]
> Add yourself to `CONTRIBUTORS.md`

thanks!

File: cli/src/aws_sso_util/populate_profiles.py (hunk 5 above)
Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py]: updated

File: cli/src/aws_sso_util/populate_profiles.py (hunk 8 above)
Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py]: heh, i just did something just like that. let me know if it works for you.

Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py]: oh, do you want me to get rid of the lambda?

Comment (benkehoe) [reply cli/src/aws_sso_util/populate_profiles.py]: yeah, I don't think the lambda is necessary. Also, I think you're missing a `.items()` for iterating over that map

Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py]: whoops! thought i'd tested the last update, but apparently not. updated again, removed the lambdas, fixed that bug.

Comment (benkehoe) [reply cli/src/aws_sso_util/populate_profiles.py]: Let's also add an `else` that raises `ValueError` for an unknown transform type. The argument processing should prevent it, but I don't like to have unhandled cases

**Review by benkehoe — APPROVED** [review]

**Comment by benkehoe** [comment]
Last thing, change the target for this pull request to be the `next` branch.

**Comment by oshaughnessy** [comment]
you got it. thanks for all the dialogue on this, ben.

File: cli/src/aws_sso_util/populate_profiles.py [hunk 9]
```diff
@@ -149,6 +149,26 @@ def trim_formatter(i, n, **kwargs):
...
     return trim_formatter
 
+def get_name_case_formatter(account_name_transform, role_name_transform, formatter):
+    field_transform_map = {
+        "account_name": account_name_transform,
+        "role_name": role_name_transform,
+    }
+    def case_formatter(i, n, **kwargs):
+        for field, transform in field_transform_map.items():
+            if transform == "capitalize":
```
Comment (benkehoe) [inline cli/src/aws_sso_util/populate_profiles.py:159]: This needs a null guard, in case only one argument is provided.
```
if not transform:
    continue
```
error:
```
aws-sso-util configure populate --dry-run --region us-east-2 --role-name-case lower
Usage: aws-sso-util configure populate [OPTIONS]

Error: Invalid profile name format: Unknown account_name transform value None
```

**Review by benkehoe — CHANGES_REQUESTED** [review]

File: cli/src/aws_sso_util/populate_profiles.py (hunk 9 above)
Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py:161]: Ooh, I saw that in the trim options and couldn't see why it was there. I assumed Click would protect us from that situation, but invoking the function for both options catches us out. Added in the most recent commit. :^)

File: cli/src/aws_sso_util/populate_profiles.py [hunk 10]
```diff
@@ -149,6 +149,28 @@ def trim_formatter(i, n, **kwargs):
...
+                continue
+            if transform == "capitalize":
+                kwargs[field] = kwargs[field].capitalize()
+            elif transform == "casefold":
+                kwargs[field] = kwargs[field].casefold()
+            elif transform == "lower":
+                kwargs[field] = kwargs[field].lower()
+            elif transform == "upper":
+                kwargs[field] = kwargs[field].upper()
+            else:
```
Comment (benkehoe) [inline cli/src/aws_sso_util/populate_profiles.py:169]: This is missing the `title` transform

Comment (oshaughnessy) [reply cli/src/aws_sso_util/populate_profiles.py:171]: added.

**Review by benkehoe — APPROVED** [review]

**Comment by benkehoe** [comment]
This has been released in [v4.27](https://github.com/benkehoe/aws-sso-util/blob/master/CHANGELOG.md#cli-v427). Thanks so much for it, it was a great idea!

**Comment by oshaughnessy** [comment]
fantastic! ✋  thanks again for engaging so much on it. i hope others find it useful as well.
